File: code-fights/arcade/03-checkPalindrome/go/palindrome.go
package main

func main() {

}

func checkPalindrome(inputString string) bool {
	tail := len(inputString) - 1
	for i := 0; i < len(inputString)/2; i++ {
		if inputString[i] != inputString[tail-i] {
			return false
		}
	}
	return true
}

// --------------------------------------------------
// FIRST SOLUTION - NAIVE
//      Do not need to loop through entire array,
//      going through half is enough.
// --------------------------------------------------
// func checkPalindrome(inputString string) bool {
//     tail := len(inputString) - 1
//     for i := range inputString {
//         if inputString[i] != inputString[tail - i] {
//             return false
//         }
//     }
//     return true
// }

File: code-fights/arcade/03-checkPalindrome/js/palindrome.js
// --------------------------------------------------
// FIRST SOLUTION - NAIVE
//      Do not need 'head' variable
// --------------------------------------------------
// function checkPalindrome(inputString) {
//     var head = 0;
//     var tail = inputString.length - 1;
//     while (head < tail) {
//         if (inputString[head] != inputString[tail]) {
//             return false
//         }
//         head++;
//         tail--;
//     }
//     return true;
// }

// // --------------------------------------------------
// COMMUNITY SOLUTION - SLOWER
//      After benchmarking, this solution proves to
//      be slightly slower.
// --------------------------------------------------
// function checkPalindrome(inputString) {
//     return inputString == inputString.split('').reverse().join('');
// }

function checkPalindrome(inputString) {
  let tail = inputString.length - 1;
  for (let i = 0; i < tail; i++) {
    if (inputString[i] !== inputString[tail]) {
      return false;
    }
    tail--;
  }
  return true;
}


checkPalindrome('racecar');
